### ISSUE TITLE
Greedy DRS: Optimization #2 

### ISSUE THREAD

**Issue by nikkolasg** [filecoin-project/drg-attacks#15]
here's an idea to improve running time of the greedy algorithm:

**Problem**: Greedy DRS exhibits a weird behavior at each iteration: the nodes selected as the top set with the highest number of incident paths, the nodes in `topk` in the code, are almost all the times getting filtered out. As a result , there is only one node to be added in the final set S at each iteration => only add one node at a time for one iteration of the attack - super slow !
The reason is once we add the *first* node from the `topk` to the set S, we update the nodes inside its radius. And since all nodes with the highest incident count are *close* to each other, then all these nodes will get filtered out! Note that this behavior is less present at the attacks is running close to the end; we see multiple nodes ending up getting selected from `topk` to the set S.

**Potential Solution**: 
An obvious solution could be to add nodes to topk "one by one" using the following dynamic strategy:
```code
Repeat until there are k elements inserted in topk {
   Pick node i with highest incident paths (and removes it from the list)
   If i is NOT in the radius set {
      add it to topk
      update the radius set from node i  (i.e. insert all nearby nodes)
  }
}
 ```

**Consequence** The iteration of the greedy attack now takes *more* time (it has to update the radius for each node added in topk, weak approximation is `O(k * d)` where `k` is the length of `topk` and `d` is the degree). At the same time the final set S grows much more *quickly* (i.e. by a `k` in fact!). The paper uses `k = 400 * 2^((n-18)/2)` i.e. with `n=2^18` we at least have `k = 400` then.  As a results, there are less nodes to analyze in the next iterations and `depth(G - S)` decreases faster.

**Comment by schomatis** [comment]
Is there an optimum radius we could look for to see if this situation improves?



> An obvious solution could be to add nodes to topk "one by one" using the following dynamic strategy:

I think I'm missing something here, that solution seems very similar to the problem description, I'm not seeing how it changes the current situation, that is, I'm not seeing much difference between the pseudo-code here and the code in `append_removal`.

**Comment by nikkolasg** [comment]
The pseudo code in the paper selects the top node depending on the count path *only*, and most of the times, the nodes will be very close by.
Here I'm saying we should select the top node depending on the count paths AND it's neighboorhood: nodes in its radius. If a node A in the radius of node B has already been selected in the topk vector, then we don't insert node B in the topk (since it will be filtered out later).

-> The big computation is this incident path counting (O(n*d)): right now, as written above, in practice, we do it once per added node in S, while it should be once for topk nodes added in S -> improving the "utility" of the big computation by a factor k (at the cost of O(k) set inclusion check)

**Comment by schomatis** [comment]
Ok, so you'd move `inradius` from `append_removal` to `count_paths`, right?

Is there a way to update `starting_paths`/`ending_paths` from the removed nodes instead of recomputing them from scratch every time?

**Comment by nikkolasg** [comment]
> Is there a way to update starting_paths/ending_paths from the removed nodes instead of recomputing them from scratch every time?

Yes ! that's a good idea: since we have the "reverse" graph (all children of a node, vs all parents of a node) as well, we can quickly update the entries of the two matrices. Although that will require them to keep them during the whole computation, leading to quite some change in the code. 
More interestingly this is one optimization that is independent of the others ! I'll try to give it a shot soon then !!

**Comment by nikkolasg** [comment]
> Ok, so you'd move inradius from append_removal to count_paths, right?

Yes exactly.

**Comment by schomatis** [comment]
On second thought, but in alignment with your initial concern, I think I would redefine the meaning of `k` (departing a bit from the paper). The main issue is that we don't want to recreate the incidents table too often (e.g., every time we remove a node).

Currently we recreate in every iteration of `greedy_reduce`. In that iteration we select the top `k` nodes for removal and later filter with radius, the problem being that we don't know how many nodes will actually be left after that (in practice it seems to be way lower than `k`).

What I would like is to repurpose `k` to mean: in this iteration remove `k` nodes. So instead of restricting the list to `topk` as in 
> `Repeat until there are k elements inserted in topk`

I would use the full list of `incidents` and keep picking from it until I have `k` to remove (this may be what you're describing already, the `topk` term throws me off a bit, it may already mean "list of `k` nodes to remove", if that is the case interpret this comment as a sign of agreement, if not, let's keep iterating).

(Independent of this, still interested in an updatable incidents mechanism.)

**Comment by nikkolasg** [comment]
That's exactly what I meant so happy we're on the same page here ! :)

**Comment by schomatis** [comment]
> I'm trying it now, the code will be a bit dirty at first, and then we can move arounds things to make it better

Take a look at https://github.com/filecoin-project/drg-attacks/pull/17/commits/48153e133ad41d51a05dcbee2e40e10fd2e9e3e9, it's pretty straightforward and you can use it as a stepping stone for further improvement.
